From 816e942c9633af90b0af24b8c32cc31659a9e351 Mon Sep 17 00:00:00 2001
From: Tomas Henzl <thenzl@redhat.com>
Date: Thu, 18 Sep 2014 14:14:24 -0400
Subject: [scsi] megaraid_sas: Round down max sge supported by controller to power of two

Message-id: <1411049668-28841-25-git-send-email-thenzl@redhat.com>
Patchwork-id: 96168
O-Subject: [RHEL7.1 e-stor PATCH 24/28] megaraid_sas : Round down max sge supported by controller to power of two
Bugzilla: 1088523
RH-Acked-by: Ewan Milne <emilne@redhat.com>
RH-Acked-by: Rob Evers <revers@redhat.com>

Round down the max sge to power of two.

Earlier max sge limit is 70 SGE, which will allow block layer to send 280K IO frame.
It is optimal to provide max IO size aligned to the smallest possible stripe size.
E.a
Consider that we have configured RAID Volumes which does not allow Fast Path across the stripe.
Raid volume with stripe size = 256K, will have peformance hit if we get io frame of size 280K.
Driver will not send IO frame large than stripe size to the Fast Path.
Also, FW will convert 280K frame into 256K + 24K. This is an additional overhead.

scsi-queue/drivers-for-3.18 commit a5fd2858e21af03555751124347e509b19c7b6f3

Signed-off-by: Sumit Saxena <sumit.saxena@avagotech.com>
Signed-off-by: Kashyap Desai <kashyap.desai@avagotech.com>
Reviewed-by: Tomas Henzl <thenzl@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/megaraid/megaraid_sas_fusion.c b/drivers/scsi/megaraid/megaraid_sas_fusion.c
index 8a27225..9eac2ef 100644
--- a/drivers/scsi/megaraid/megaraid_sas_fusion.c
+++ b/drivers/scsi/megaraid/megaraid_sas_fusion.c
@@ -1001,8 +1001,8 @@ megasas_init_adapter_fusion(struct megasas_instance *instance)
  fusion->max_sge_in_chain =
   MEGASAS_MAX_SZ_CHAIN_FRAME / sizeof(union MPI2_SGE_IO_UNION);
 
- instance->max_num_sge = fusion->max_sge_in_main_msg +
-  fusion->max_sge_in_chain - 2;
+ instance->max_num_sge = rounddown_pow_of_two(
+  fusion->max_sge_in_main_msg + fusion->max_sge_in_chain - 2);
 
  /* Used for pass thru MFI frame (DCMD) */
  fusion->chain_offset_mfi_pthru =
-- 
1.7.1